Image block: fix replace link control styling

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

The DOM and styling updates to `<LinkControl>` necessitated by the introduction of the the rich link previews feature caused the Image block `Replace` flow dropdown to suffer a visual layout regression.

![123964709-c82ad600-d9c4-11eb-87f0-9378698b1c3f](https://user-images.githubusercontent.com/444434/125062061-e91ab780-e0a5-11eb-86a0-eee8704ff4d4.png)

This PR resolves this by aligning everything nicely again. See screenshots below.

Fixes https://github.com/WordPress/gutenberg/issues/33096

## How has this been tested?

* Add an Image block.
* Upload an Image and set it.
* Click `Replace` on the Image block.
* See the link UI without any obvious visual layout glitches.

## Screenshots <!-- if applicable -->

| Before | After |
| ------ | ------ | 
| ![123964709-c82ad600-d9c4-11eb-87f0-9378698b1c3f](https://user-images.githubusercontent.com/444434/125062061-e91ab780-e0a5-11eb-86a0-eee8704ff4d4.png) | <img width="445" alt="Screenshot 2021-07-09 at 11 08 49" src="https://user-images.githubusercontent.com/444434/125062217-0f405780-e0a6-11eb-87f0-e62a669da914.png"> |

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Conditionalise display of wrapping “tools” div

getdave committed:
Align items visually to top of control below “Current media URL:” text

getdave commented:
Fixing those tests now...

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -261,23 +266,26 @@ function LinkControl( {
-					</Button>
-				) }
-			</div>
+			{ showSettingsDrawer ||
```
Whoops! This introduces a regression. Thank you unit tests (which I wrote 😉 )!

So...let me think about this.

getdave replied:
Fixed.

getdave committed:
Extract default settings to root component scope to ensure tools div renders by default

This preserves the existing functionality of the control.

getdave commented:
@javierarce a quick 👀 from a design perspective would be good here if you have any time.

javierarce commented:
This looks much better, @getdave. 

I just saw a couple of things:

![image](https://user-images.githubusercontent.com/4933/125068659-03f52800-e0b6-11eb-9893-44b64a8e82ef.png)

- I think we should have the same spacing between the `Current media URL` label and the input field. 8px would be perfect. 
(I also noticed that there's a weird 1px padding around the `.block-editor-url-input` element.)
- If I edit the URL the box shrinks (I understand this could difficult to fix).
- We are showing the URL of the image twice, which is not very useful. Could we have something similar to the previous component?

<img width="264" alt="Screenshot 2021-07-09 at 13 23 03" src="https://user-images.githubusercontent.com/4933/125070887-d52c8100-e0b8-11eb-8792-26f0b96595dd.png">

I also discovered that if you enter a Base64 encoded image, the URL that appears in the `.block-editor-link-control__search-item-info` doesn't get shortened like other URLs. This is probably an extreme case, but I just thought it was worth mentioning. 

<img width="748" alt="Screenshot 2021-07-09 at 13 11 12" src="https://user-images.githubusercontent.com/4933/125069539-2471b200-e0b7-11eb-89d1-f3147a349966.png">

draganescu reviewed (approved):
This works as described. Code looks good. 👍🏻

getdave committed:
Add space between label and input

As requested by designer in https://github.com/WordPress/gutenberg/pull/33326#issuecomment-877119583

getdave committed:
Fix popover resizing when switching between edit and preview mode

getdave commented:
Thanks for the review @javierarce!

> * I think we should have the same spacing between the `Current media URL` label and the input field. 8px would be perfect.

I've added a consistent 8px margin in both edit and preview mode.

>   (I also noticed that there's a weird 1px padding around the `.block-editor-url-input` element.)

Removed that.

> * If I edit the URL the box shrinks (I understand this could difficult to fix).

I set a hardcoded 200px width which avoids this. 

> * We are showing the URL of the image twice, which is not very useful. Could we have something similar to the previous component?

I don't think we should attempt to "fix" this here but in a follow up. Essentially we do have much more data available but it's not being passed to the component so it isn't being displayed.

<img width="1261" alt="Screen Shot 2021-07-09 at 14 35 16" src="https://user-images.githubusercontent.com/444434/125087524-766d0480-e0c4-11eb-876b-337d99e27b13.png">


> I also discovered that if you enter a Base64 encoded image, the URL that appears in the `.block-editor-link-control__search-item-info` doesn't get shortened like other URLs. This is probably an extreme case, but I just thought it was worth mentioning.

I think that should be resolved by the fixed width above.

getdave commented:
Rebasing FTW

javierarce commented:
The changes look good to me! 🚀

> I don't think we should attempt to "fix" this here but in a follow up. Essentially we do have much more data available but it's not being passed to the component so it isn't being displayed.

Ok!

Mamaduka reviewed (approved):
LGTM :shipit:

Thanks for working on this, @getdave.

getdave commented:
Tests failing probably needs a rebase. Please could someone merge this so it makes 11.1 RC?

Mamaduka commented:
@getdave, all failing tests should be skipped soon. Maybe one more rebase after #33353.

JustinyAhin commented:
@getdave this works as expected. Fine for me :).

Thanks

fullofcaffeine reviewed (approved):
This tests well! 🚢

getdave commented:
Tests finally passed. I didn't do anything so I guess these were upstream issues on `trunk`.

getdave commented:
@youknowriad This might be a good one for cherry picking in `11.1.0`.

cc @gwwar who I think is shadowing the release.

youknowriad commented:
Changed the milestone to consider cherry-picking this.

getdave commented:
Thanks @youknowriad. Wasn't sure whether to go ahead and do that without confirming with the release lead first.
